### ISSUE TITLE
how come () => void is a subtype of () => a if void isn't a subtype of a and a isn't subtype of void?

### ISSUE THREAD

**Issue by aleksey-bykov** [Microsoft/TypeScript#8581]
```typescript

declare var a: () => void;
declare var b: () => PromiseLike<void>;

a = b; // <-- why!!!
b = a;


declare var c: void;
declare var d: PromiseLike<void>;

c = d;
d = c;


```

![image](https://cloud.githubusercontent.com/assets/937933/15227595/f9ab80ac-1855-11e6-9cac-66d09dcc75b3.png)

**Comment by vladima** [comment]
*tl;dr;* the assumption is that it is save to assign something that returns some value to something that returns `void` so result is ignored, however the opposite is not true.

see [3.11.3 Subtypes and Supertypes](https://github.com/Microsoft/TypeScript/blob/master/doc/spec.md#3113-subtypes-and-supertypes)

>* *S* is an object type, an intersection type, an enum type, or the Number, Boolean, or String primitive type, *T* is an object type, and for each member *M* in *T*, one of the following is true:
  * *M* is a property and *S* has an apparent property *N* where
    * *M* and *N* have the same name,
    * the type of *N* is a subtype of that of *M*,
    * if *M* is a required property, *N* is also a required property, and
    * *M* and *N* are both public, *M* and *N* are both private and originate in the same declaration, *M* and *N* are both protected and originate in the same declaration, or *M* is protected and *N* is declared in a class derived from the class in which *M* is declared.
  * *M* is a non-specialized call or construct signature and *S* has an apparent call or construct signature *N* where, when *M* and *N* are instantiated using type Any as the type argument for all type parameters declared by *M* and *N* (if any),
    * the signatures are of the same kind (call or construct),
    * *M* has a rest parameter or the number of non-optional parameters in *N* is less than or equal to the total number of parameters in *M*,
    * for parameter positions that are present in both signatures, each parameter type in *N* is a subtype or supertype of the corresponding parameter type in *M*, and
    * the result type of *M* is **Void**, or the result type of *N* is a subtype of that of *M*.

**Comment by aleksey-bykov** [comment]
it's not a safe assumption at all, ugh....

**Comment by aleksey-bykov** [comment]
I have a function that returns a promise: `() => Promise<void>, it is passed around a callback, after refactoring unintentionally it end up being an argument of a parameter of type `() => void`. Right here I need the compiler's help to notice it and tell me that I can't use it there, because `Promise<void>` just can't be treated the same way as `void`, and what does compiler do? nothing, exactly

**Comment by aleksey-bykov** [comment]
anything that has to be serialized can't be passed round safely because there is a chance it will get swallowed by `() => void` and then end up being passed to the standard `JSON.stringify` which will then crash

**Comment by RyanCavanaugh** [comment]
The alternative is much worse -- imagine not being able to write this code:
```ts
let arr = [1, 2];
// Error, can't convert 'number' return type to 'void'
[1, 2, 3].forEach(x => arr.push(x));
```

**Comment by aleksey-bykov** [comment]
much worse, I agree, you sound more convincing than never

```
let arr = [1, 2];
// Error, can't convert 'number' return type to 'void'
[1, 2, 3].forEach(x => { arr.push(x); });
```

**Comment by RyanCavanaugh** [comment]
If that's sarcasm I'm missing it :stuck_out_tongue:

**Comment by aleksey-bykov** [comment]
not much sarcasm in 3 extra characters
